hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -317,9 +335,33 @@ private TransactionBody handleUserTransaction(
             } else {
                 feeAccumulator.charge(payer, fees);
                 try {
+                    if (!isPayerThrottleExempt && networkUtilizationManager.wasLastTxnGasThrottled()) {
+                        // Refund the service fees already charged to the payer, because the user-submitted transaction
+                        // was fully valid but network capacity was unavailable to satisfy it.
+                        final var serviceFee = new Fees(0L, 0L, fees.serviceFee());
+                        feeAccumulator.refund(payer, serviceFee);
+                        throw new HandleException(CONSENSUS_GAS_EXHAUSTED);
+                    }
```

Comment:
Can't we just move that before we charge the user(line 336) so we don't charge the user in the first place? Or do that so we can have an event history of some sort?

Reply:
I implemented it that way because that's how it's done in mono.
If I move it up I would need to have a conditional and for this case we will need to charge the node and network components only and otherwise charge all three components i.e.
```
try {
       if (!isPayerThrottleExempt && networkUtilizationManager.wasLastTxnGasThrottled()) {
              final var nodeAndNetworkFees = new Fees(fees.nodeFee(), fees.networkFee(), 0L);
              feeAccumulator.charge(payer, nodeAndNetworkFees);
              throw new HandleException(CONSENSUS_GAS_EXHAUSTED);
       } else {
              feeAccumulator.charge(payer, fees);
       }
...
```
It doesn't look much different to me, not sure what we gain if I do it that way?

Reply:
Oh, the fees are different. Gotcha. I thought that the fees were the same and that we were charging the user with a fee and then if the last transaction was gas throttled - we just refunded the same fee.